File: backend/src/entities/GeneratedFile.ts
import { Entity, PrimaryGeneratedColumn, Column } from "typeorm";

export type GeneratedFileType = "html" | "js" | "css" | "data-js";

/** One generated output file, owned by either an Upload (Excel-authored) or a
 * FormVersion (builder-authored) — exactly one of `uploadId`/`formVersionId` is set,
 * enforced by a CHECK constraint (see the AddFormVersionToGeneratedFiles migration),
 * not by TypeScript — a plain `@Column` pair remains the simplest way to model "one of
 * two nullable owner columns" in this codebase's existing entity style. */
@Entity("GeneratedFiles")
export class GeneratedFile {
  @PrimaryGeneratedColumn("uuid")
  id!: string;

  @Column({ type: "uniqueidentifier", nullable: true })
  uploadId!: string | null;

  @Column({ type: "uniqueidentifier", nullable: true })
  formVersionId!: string | null;

  @Column({ type: "nvarchar", length: 255 })
  fileName!: string;

  @Column({ type: "nvarchar", length: 2000 })
  filePath!: string;

  @Column({ type: "nvarchar", length: 20 })
  fileType!: GeneratedFileType;

  @Column({ type: "datetimeoffset", default: () => "SYSDATETIMEOFFSET()" })
  createdAt!: Date;
}

File: backend-py/app/models/generated_file.py
"""Port of `backend/src/entities/GeneratedFile.ts`.

One generated output file, owned by a `FormVersion` (`uploadId` is a leftover
nullable column from the removed Excel-upload feature — never populated by
any surviving code path; `formVersionId` is the real owner now).
"""

from __future__ import annotations

from typing import Literal, Optional

from sqlalchemy.orm import Mapped

from app.models.base import Base, datetimeoffset_now, nvarchar, uuid_col, uuid_pk

GeneratedFileType = Literal["html", "js", "css", "data-js"]


class GeneratedFile(Base):
    __tablename__ = "GeneratedFiles"

    id: Mapped[str] = uuid_pk()
    uploadId: Mapped[Optional[str]] = uuid_col(nullable=True)
    formVersionId: Mapped[Optional[str]] = uuid_col(nullable=True)
    fileName: Mapped[str] = nvarchar(255)
    filePath: Mapped[str] = nvarchar(2000)
    fileType: Mapped[GeneratedFileType] = nvarchar(20)
    createdAt = datetimeoffset_now()
